pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterMetricsStats.java

```diff
@@ -0,0 +1,234 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.transaction.coordinator.impl;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Histogram;
+import java.io.Closeable;
+import java.util.HashMap;
+
+/***
+ * Describes the working status of the {@link TxnLogBufferedWriter}, helps users tune the thresholds of
+ * {@link TxnLogBufferedWriter} for best performance.
+ * Note-1: When batch feature is turned off, no data is logged at this. In this scenario，users can see the
+ *    {@link org.apache.bookkeeper.mledger.ManagedLedgerMXBean}.
+ * Note-2: Even if enable batch feature, if a single record is too big, it still directly write to Bookie without batch,
+ *    property {@link #pulsarBatchedLogTriggeringCountByForce} can indicate this case. But this case will not affect
+ *    other metrics, because it would obscure the real situation. E.g. there has two record:
+ *    [{recordsCount=512, triggerByMaxRecordCount}, {recordCount=1, triggerByTooLarge}], we should not tell the users
+ *    that the average batch records is 256, if the users knows that there are only 256 records per batch, then users
+ *    will try to increase {@link TxnLogBufferedWriter#batchedWriteMaxDelayInMillis} so that there is more data per
+ *    batch to improve throughput, but that does not work.
+ */
+public class TxnLogBufferedWriterMetricsStats implements Closeable {
+
+    /**
+     * Key is the name we used to create {@link TxnLogBufferedWriterMetricsStats}, and now there are two kinds:
+     * ["pulsar_txn_tc_log", "pulsar_txn_tc_batched_log"]. There can be multiple labels in each
+     * {@link TxnLogBufferedWriterMetricsStats}, such as The Transaction Coordinator using coordinatorId as label and
+     * The Transaction Pending Ack Store using subscriptionName as label.
+     */
+    private static final HashMap<String, Collector> COLLECTOR_CACHE = new HashMap<>();
```

Comment:
Why do we need a map here?
Looks like it only used in the constructor.

Reply:
> Why do we need a map here? Looks like it only used in the constructor.

Yes, we need.

To build MetricsStat, we need execute these two steps:
1. Create `Collector` and register to `CollectorRegistry`, perhaps the Collector is `Histogram` or `Counter` 
2. Register labels to `Collector` and get `Collector.child`. This step can also be omitted, because we can execute `collector.labels(labelValues)` to get it. 

In the Transaction log scenario, multiple Transaction Log share the same `Collector`, and each has its own `Collector.Child`, so when we build metrics stat for each Transaction Log, we need to get the `Collector` to get the `Collector.Child`. However, the CollectorRegistry does not provide an API like `collectorRegistry.getRegistedCollector(String name)`, and it will throws IllegalArgumentException when we registering collector with the same name more than once, see:


https://github.com/prometheus/client_java/blob/1966186deaaa83aec496d88ff604a90795bad688/simpleclient/src/main/java/io/prometheus/client/CollectorRegistry.java#L49-L65

So we have to manage the registered collectors ourselves.

Reply:
That's a good way to do it, but this can make the code confuse:

- When the Transaction Log Provider opens a Transaction Log, passed the `Histogram` to Transaction Log. That is OK.
- When the Transaction Log close, remove the labels. That is ok too.

But the Transaction Log Provider will hold all the `Collector` of Txn Buffered Writer, this is confusing

Reply:
Hi @asafm 

> Can you explain this further? I didn't understand how the provider would hold all the Collectors (metrics). It should only be to a single instance of the metric class.

I responded separately below

Reply:
Yes, maybe more than one.

Reply:
> Ok, if more than one, then the design must change. I thought the whole idea was that you have a single instance of metrics per metric prefix.
> If not, after much thought I suggest the following:

```
abstract class BufferMetrics {
	protected abstract void observeRecordsPerBatch(int)
	protected abstract void incFlushTriggeredByMaxRecords(int)
}

MLTransactionMetadataStoreBufferedWriterMetrics extends BufferMetrics {
	static private Histogram recordsPerBatchMetric = new Histogram.Builder()
                        .name("pulsar_tx_store_bufferedwriter_batch_record_count")
                        .labelNames(new String[]{"txCoordinatorId"})
                        .help("Records per batch histogram")
                        .buckets(RECORD_COUNT_PER_ENTRY_BUCKETS)
                        .register(registry));
	
    private Histogram.Child recordsPerBatchHistogram;                      


	public MLTransactionMetadataStoreBufferedWriterMetrics(String txCoordinatorId) {
	    recordsPerBatchHistogram = recordsPerBatchHistogram.labels(txCoordinatorId)

	}
     
       protected observeRecordsPerBatch(value) {
            recordsPerBatchHistogram.observe(value)       
       }

}
```

> Another approach which I disliked a bit, but it's still ok:
> Add to Pulsar Common:

```
class PrometheusRegistryChecker {
     static defaultMetricRegistryNameToCollector = new HashMap<String, Collector>()

     static Collector registerIfNotExists(collector) {}
}
```

> Like FunctionCollectorRegistryImpl

I prefer the second implementation because if we need to define many variables of Collector type. And `Map` is really just another representation of multiple variables.

Now `common` has not the dependency: `prometheus-client`